rclpy/src/rclpy/_rclpy.c

```diff
@@ -231,6 +231,307 @@ rclpy_create_subscription(PyObject * Py_UNUSED(self), PyObject * args)
+  PyList_Append(pylist, pyclient);
+  PyList_Append(pylist, PyLong_FromUnsignedLongLong((uint64_t)&client->impl));
+
+  return pylist;
+}
+
+/// Publish a request message
+/*
+ * \param[in] pyclient Capsule pointing to the client
+ * \param[in] pyrequest request message to send
+ * \return sequence_number PyLong object representing the index of the sent request
+ */
+static PyObject *
+rclpy_send_request(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pyclient;
+  PyObject * pyrequest;
+
+  if (!PyArg_ParseTuple(args, "OO", &pyclient, &pyrequest)) {
+    return NULL;
+  }
+  rcl_client_t * client = (rcl_client_t *)PyCapsule_GetPointer(pyclient, NULL);
+  assert(client != NULL);
+
+  PyObject * pyrequest_type = PyObject_GetAttrString(pyrequest, "__class__");
+  assert(pyrequest_type != NULL);
+
+  PyObject * pymetaclass = PyObject_GetAttrString(pyrequest_type, "__class__");
+  assert(pymetaclass != NULL);
+
+  PyObject * pyconvert_from_py = PyObject_GetAttrString(pymetaclass, "_CONVERT_FROM_PY");
+
+  assert(pyconvert_from_py != NULL);
+  typedef void * (* convert_from_py_signature)(PyObject *);
+  convert_from_py_signature convert_from_py =
+    (convert_from_py_signature)PyCapsule_GetPointer(pyconvert_from_py, NULL);
+
+  assert(convert_from_py != NULL);
+  void * raw_ros_request = convert_from_py(pyrequest);
+  int64_t sequence_number;
+  rcl_ret_t ret = rcl_send_request(client, raw_ros_request, &sequence_number);
+  if (ret != RCL_RET_OK) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to send request: %s", rcl_get_error_string_safe());
+    return NULL;
+  }
+
+  return PyLong_FromLongLong(sequence_number);
+}
+
+/// Create a service server
+/*
+ * This function will create a service server and attach it to the topics <pyservice_name>Request and <pyservice_name>Reply
+ * This service will use the typesupport defined in the service module
+ * provided as pysrv_type to send messages over the wire.
+ *
+ * \param[in] pynode Capsule pointing to the node to add the service to
+ * \param[in] pysrv_type Service module associated with the service
+ * \param[in] pyservice_name Python object containing the name used to generate the
+ *   <pyservice_name>Request and <pyservice_name>Reply topics names
+ * \param[in] pyqos_profile QoSProfile Python object with the profile of this service
+ * \return NULL on failure
+ *         List with 2 elements:
+ *            first element: a Capsule pointing to the pointer of the created rcl_service_t * structure
+ *            second element: an integer representing the memory address of the created rcl_service_t
+ */
+static PyObject *
+rclpy_create_service(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pynode;
+  PyObject * pysrv_type;
+  PyObject * pyservice_name;
+  PyObject * pyqos_profile;
+
+  if (!PyArg_ParseTuple(args, "OOOO", &pynode, &pysrv_type, &pyservice_name, &pyqos_profile)) {
+    return NULL;
+  }
+
+  assert(PyUnicode_Check(pyservice_name));
```

Comment:
While some of the asserts are checks for internal use this one (and others, e.g. https://github.com/ros2/rclpy/pull/51/files#diff-80dda03110b5be606b823d6b0558b5c2R372) stood out to me since it depends on user provided arguments and can absolutely fail in a release build. I don't think the existing ones (as far as I have seen) need to be updated. But the new ones which are dealing with user provided data should be imo.